### PR TITLE
GDB stub fixes

### PR DESCRIPTION
**Pull request by WinterMute**
These changes allow the gdb stub to be enabled and configured from the command line which allows for easier integration with IDEs that do source level debugging.

GDB complains Remote 'g' packet reply is too long with this setup because citra is attempting to send float registers it's not expecting, even if I strip out the legacy fpa registers which are no longer used. I'm still attempting to figure out how to get gdb to accept vfp registers.

The Qt build also appears to have problems with gdb communication but these patches don't change that behaviour.

### CODE REVIEW

**Comment by neobrain** [comment]
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-01-30T14:32:12Z: https://github.com/citra-emu/citra/compare/3ab6d77c206c33e5036fcde1de0663c2f72777ea...WinterMute:9c4d9f4aa5c406d57c4f7ffb6438ab8d47de7d52

**Comment by jroweboy** [comment]
@WinterMute You can tell GDB the list of registers as one of the query parameters. For example, I used this target description https://github.com/jroweboy/citra/blob/gdb-stub/src/core/arm/gdb_stub.cpp#L41 which should get sent in response to gdb's query packet. https://github.com/jroweboy/citra/blob/gdb-stub/src/core/arm/gdb_stub.cpp#L251 for where I sent it, and it should match up to citra's gdb stub here https://github.com/citra-emu/citra/blob/master/src/core/gdbstub/gdbstub.cpp#L353

Note that according to the documentation, one should be able to just tell GDB to use the target description from its source, but I could never seem to get that to work https://sourceware.org/gdb/onlinedocs/gdb/Target-Descriptions.html#Target-Descriptions

**Comment by neobrain** [comment]
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-02-01T13:10:29Z: https://github.com/citra-emu/citra/compare/3ab6d77c206c33e5036fcde1de0663c2f72777ea...WinterMute:c99a4079980822255f7c6ee5e7ba99957c741aaa

File: src/citra/citra.cpp [hunk 1]
```diff
@@ -34,25 +34,40 @@
...
 /// Application entry point
 int main(int argc, char **argv) {
     int option_index = 0;
+    u32 gdb_port = 0;
+    char *endarg;
     std::string boot_filename;
+
     static struct option long_options[] = {
         { "help", no_argument, 0, 'h' },
+        { "gdbport", required_argument, 0, 'g' },
```
Comment (Subv) [inline src/citra/citra.cpp]: Why is this required?

File: src/core/gdbstub/gdbstub.cpp [hunk 2]
```diff
@@ -353,8 +406,15 @@ static void SendReply(const char* reply) {
...
 
-    if (!strcmp(reinterpret_cast<const char*>(command_buffer + 1), "TStatus")) {
+    const char *query = reinterpret_cast<const char*>(command_buffer + 1);
+
+    if (strcmp(query, "TStatus") == 0 ) {
         SendReply("T0");
+    } else if (strncmp(query, "Supported:",strlen("Supported:")) == 0) {
+        /// PacketSize needs to be large enough for target xml
+        SendReply("PacketSize=800;qXfer:features:read+");
+    } else if (strncmp(query, "Xfer:features:read:target.xml:",strlen("Xfer:features:read:target.xml:")) == 0) {
```
Comment (Subv) [inline src/core/gdbstub/gdbstub.cpp]: Whitespace before each parameter

**Comment by neobrain** [comment]
@WinterMute While you're at fixing up stuff, I also suggest setting SO_REUSEADDR on the stub socket, since otherwise connecting to the stub will fail every second Citra run or so.

**Comment by bunnei** [comment]
@WinterMute do you plan on responding to the above comments?

**Comment by WinterMute** [comment]
@bunnei yeah, sorry. Haven't got around to looking at this again yet. Will do soon

File: src/core/gdbstub/gdbstub.cpp [hunk 3]
```diff
@@ -353,8 +406,15 @@ static void SendReply(const char* reply) {
 static void HandleQuery() {
     LOG_DEBUG(Debug_GDBStub, "gdb: query '%s'\n", command_buffer + 1);
 
-    if (!strcmp(reinterpret_cast<const char*>(command_buffer + 1), "TStatus")) {
+    const char *query = reinterpret_cast<const char*>(command_buffer + 1);
```
Comment (bunnei) [inline src/core/gdbstub/gdbstub.cpp]: `const char* query` (minor whitespace nitpick)

File: src/core/gdbstub/gdbstub.cpp [hunk 4]
```diff
@@ -353,8 +406,15 @@ static void SendReply(const char* reply) {
...
+
+    if (strcmp(query, "TStatus") == 0 ) {
         SendReply("T0");
+    } else if (strncmp(query, "Supported:",strlen("Supported:")) == 0) {
```
Comment (bunnei) [inline src/core/gdbstub/gdbstub.cpp]: whitepace

File: src/core/gdbstub/gdbstub.cpp [hunk 5]
```diff
@@ -353,8 +406,15 @@ static void SendReply(const char* reply) {
...
+        /// PacketSize needs to be large enough for target xml
```
Comment (bunnei) [inline src/core/gdbstub/gdbstub.cpp]: `//`

Comment (bunnei) [inline src/core/gdbstub/gdbstub.cpp]: XML

File: src/core/gdbstub/gdbstub.cpp [hunk 6]
```diff
@@ -60,6 +60,59 @@ const u32 R15_REGISTER = 15;
 const u32 CPSR_REGISTER = 25;
 const u32 FPSCR_REGISTER = 58;
 
+// for sample xmls see the gdb source /gdb/features
```
Comment (bunnei) [inline src/core/gdbstub/gdbstub.cpp]: Super huge pedantic nit, but I'm going to be _that guy_ anyways:
<pre>
// For sample XML fikes see the GDB source /gdb/features
// GDB also wants the l character at the start
// This XML defines what the registers are for this specific ARM device
</pre>

**Comment by bunnei** [comment]
Superceded by #1644, @polaris- offered to wrap this one up!
